New connection manager events, ConnectionManagerRequestService

bmen25124 opened the pull request:
- New events for connection profiles: `CONNECTION_PROFILE_CREATED`, `CONNECTION_PROFILE_DELETED`, `CONNECTION_PROFILE_UPDATED`
- Added dompurify to profile names.
- Added `presetToGeneratePayload` methods to `custom-request.js` 
- Added instruct template support for text generation in `custom-request.js` 
- Added reasoning extract to `custom-request.js`

Reithan committed:
continue works same as swipe continued message isn't depth counted

Reithan committed:
correct early-out check

bmen25124 committed:
Addded connection profile support for 'memory' extension, CONNECTION_PROFILE_DELETED/CONNECTION_PROFILE_CREATED events

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +644,187 @@ async function getSummaryPromptForNow(context, force) {
+    /**
+     * @type {number}
+     */
+    const adjustedMaxTokens = extension_settings.memory.overrideResponseLength || 4096;
```
We need to reconsider a 4096 as a fallback since it doesn't seem a very safe default.
Maybe 2048 would be a bit better.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +644,187 @@ async function getSummaryPromptForNow(context, force) {
+ * @param {string} prompt Prompt text
+ * @returns {Promise<string|null>} Generated text or null if an error occurred
+ */
+async function getGeneratePayloadWithProfile(profileId, prompt) {
```
I think it's better to be moved to `scripts/extensions/shared.js` as it seems pretty reusable. See WebLLM methods there.

bmen25124 replied:
I moved to `shared.js`. I added a couple of methods... to create a foundation for `connection manager - custom request`

bmen25124 committed:
Added ConnectionManagerUtils class for extensions,  added CONNECTION_PROFILE_UPDATED event

bmen25124 committed:
ConnectionManagerUtils -> ConnectionManagerRequestService

Wolfsblvt commented on public/scripts/extensions/memory/settings.html:
```diff
@@ -14,8 +14,25 @@
+                <div data-summary-source="profile" style="display:none;">
+                    <div class="openai_restorable flex-container">
+                        <div class="title_restorable">
+                            <label>Connection Profile</label>
```
```suggestion
                            <label data-i18n="Connection Profile">Connection Profile</label>
```
Help our translator friends.

Wolfsblvt commented on public/scripts/extensions/memory/settings.html:
```diff
@@ -14,8 +14,25 @@
+                            <label>Connection Profile</label>
+                            <button
+                                class="menu_button fa-solid fa-route interactable redirect_sys_settings"
+                                title="Go to connection profile tab"
```
```suggestion
                                title="Go to connection profile tab" data-i18n="[title]Go to connection profile tab"
```

Wolfsblvt commented on public/scripts/extensions/memory/settings.html:
```diff
@@ -14,8 +14,25 @@
+                            ></button>
+                        </div>
+                        <select id="memory_connection_profile" class="text_pole profile" title="Select a Connection Profile">
+                            <option value="">Select a Connection Profile</option>
```
```suggestion
                            <option value="" data-i18n="Select a Connection Profile">Select a Connection Profile</option>
```

Wolfsblvt commented on public/scripts/extensions/memory/index.js:
```diff
@@ -1047,6 +1199,10 @@ jQuery(async function () {
     await addExtensionControls();
     loadSettings();
+
+    eventSource.on(event_types.CONNECTION_PROFILE_CREATED, (profile) =>populateProfileDropdown('create', profile));
```
```suggestion
    eventSource.on(event_types.CONNECTION_PROFILE_CREATED, (profile) => populateProfileDropdown('create', profile));
```
linting

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +646,119 @@ async function getSummaryPromptForNow(context, force) {
+
+    if (type === 'refresh') {
+        dropdown.empty();
+        dropdown.append('<option value="">Select a Connection Profile</option>');
```
The text is not localizable. Might be good to add i18n attribute.

Wolfsblvt replied:
i18n here too (should work with the MutationObserver, or simply use t``.

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +646,119 @@ async function getSummaryPromptForNow(context, force) {
+        }
+    } else if (type === 'create' && firstProfile) {
+        if (ConnectionManagerRequestService.isProfileSupported(firstProfile)) {
+            dropdown.append(`<option value="${firstProfile.id}">${firstProfile.name}</option>`);
```
Avoid using strings for inserting elements. Prefer document.createElements. This code is very vulnerable to HTML injections.

bmen25124 replied:
Fixed but I don't understand the problem. Where is the vulnerability exactly 🤭

Wolfsblvt replied:
Name your connection profile `Heh<script>alert('Möööööp')</script>` with the code you had before.
Extension profiles are exportable. We should at least somewhat check for sanitization.

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +646,119 @@ async function getSummaryPromptForNow(context, force) {
+    let index = null;
+
+    if (extension_settings.memory.prompt_builder === prompt_builders.DEFAULT) {
+        toastr.info('Connection profile does\'t support this prompt builder. Using raw prompt.');
```
I think message is better to be silent

Wolfsblvt replied:
Or just debug log it to console. Doesn't need to be visible for the user.

bmen25124 replied:
Fixed, converted to debug

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +646,119 @@ async function getSummaryPromptForNow(context, force) {
+    const newContext = getContext();
+
+    // something changed during summarization request
+    if (newContext.groupId !== context.groupId
```
I made a shared func for this logic. Please look it up in this module. Thanks.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/memory/settings.html:
```diff
@@ -14,8 +14,25 @@
                 </select><br>
 
+                <!-- Connection Profile section (only visible when "profile" is selected) -->
+                <div data-summary-source="profile" style="display:none;">
```
I don't think you need display:none here, this is set automatically by the extension code on init.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +269,125 @@ export async function getWebLlmContextSize() {
 }
+
+/**
+ * Doesn't support streaming.
```
Class summary seems lacking to me

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +269,125 @@ export async function getWebLlmContextSize() {
+ */
+export class ConnectionManagerRequestService {
+    static ALLOWED_TYPES = {
+        openai: 'Chat Completion',
```
Use localized names (t function from i18n.js)

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +269,125 @@ export async function getWebLlmContextSize() {
+        }
+
+        try {
+            if (selectedApiMap.selected === 'openai') {
```
I think you can extract this if/else logic to private functions and use switch to select a one that works for the selected API.

bmen25124 replied:
I tried, when I add new method, it became cluster fuck because there are specific params text completion and chat completion. I think we should leave it.

bmen25124 committed:
Added handleDropdown

bmen25124 committed:
Added i18n, using document.createElement instead of string insertion, better prompt format

bmen25124 committed:
Removed unencessary empty <option> which already add in code, removed index access

bmen25124 commented:
Added i18n attributes to all HTML/some JS elements.

bmen25124 committed:
Added connectionManager check

Wolfsblvt commented on public/scripts/extensions/shared.js:
```diff
@@ -424,16 +431,26 @@ export class ConnectionManagerRequestService {
+        const defaultOption = document.createElement('option');
+        defaultOption.value = '';
+        defaultOption.textContent = 'Select a Connection Profile';
+        defaultOption.setAttribute('data-i18n', 'Select a Connection Profile');
```
That's okay. But when not using jQuery (thank god), there is a neat and easy thing to set data attributes.
```suggestion
        defaultOption.dataset.i18n = 'Select a Connection Profile';
```

bmen25124 replied:
Fixed

Wolfsblvt reviewed (commented):
Good on the side of translations and general code, from what I can see.

Haven't tested this though, and goes a bit too deep into the internal stuff. You guys can play this out.
Good for me.

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +654,79 @@ async function getSummaryPromptForNow(context, force) {
+        console.debug('Connection profile does\'t support this prompt builder. Using raw prompt.');
+    }
+
+    const lock = extension_settings.memory.prompt_builder === [prompt_builders.RAW_BLOCKING, prompt_builders.DEFAULT].includes(extension_settings.memory.prompt_builder);
```
What's going on this line? You're comparing strings with boolean with strict type check.

Should it be just? 
```js
const lock = [prompt_builders.RAW_BLOCKING, prompt_builders.DEFAULT].includes(extension_settings.memory.prompt_builder);
```

bmen25124 replied:
I might be high, fixed

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +654,79 @@ async function getSummaryPromptForNow(context, force) {
+    const newContext = getContext();
+
+    // something changed during summarization request
+    if (isContextChanged(newContext)) {
```
`isContextChanged` creates a new context inside. You should call getContext BEFORE calling to make a summary and then pass a reference to `isContextChanged`

bmen25124 replied:
Fixed. Sorry, I'm trying to replicate what WebLLM does, which I failed in this case.

bmen25124 committed:
Added sanitization for profile name, fixed lock variable, setAttribute -> dataset

bmen25124 committed:
Fixed name diff check

bmen25124 committed:
Fixed isContextChanged param

Wolfsblvt commented on public/scripts/extensions/connection-manager/index.js:
```diff
@@ -555,11 +557,11 @@ async function renderDetailsContent(detailsContent) {
         }
 
-        if (profile.name !== newName && extension_settings.connectionManager.profiles.some(p => p.name === newName)) {
+        if (profile.name !== userInput && extension_settings.connectionManager.profiles.some(p => p.name === userInput)) {
```
If you purify the profiles that you save, you need to purify on rename too, before making those checks. This, and the one after.
Otherwise the resulting profile name *could* be the same as a profile that already exists, but this error is not triggering.

bmen25124 replied:
I fixed it after this commit 👽

bmen25124 committed:
Changed variable name to before

bmen25124 committed:
Better error type

Cohee1207 commented on public/scripts/extensions/connection-manager/index.js:
```diff
@@ -254,8 +254,8 @@ async function createConnectionProfile(forceName = null) {
-    const name = forceName ?? await callGenericPopup(template, POPUP_TYPE.INPUT, suggestedName, { rows: 2 });
-
+    let name = forceName ?? await callGenericPopup(template, POPUP_TYPE.INPUT, suggestedName, { rows: 2 });
+    name = DOMPurify.sanitize(String(name));
```
If you cancel the popup of INPUT type, the result will be recorded as "false". Please move sanitation after the if (!name) check.

bmen25124 replied:
Fixed. New code:
```js
    let name = forceName ?? await callGenericPopup(template, POPUP_TYPE.INPUT, suggestedName, { rows: 2 });
    // If it's cancelled, it will be false
    if (!name) {
        return null;
    }
    name = DOMPurify.sanitize(String(name));
    if (!name) {
        toastr.error('Name cannot be empty.');
        return null;
    }
   ```
   
I still check if it's empty. I'm not sure it is necessary or not. You decide.

Cohee1207 replied:
> I'm not sure it is necessary

It is. thanks.

Cohee1207 commented on public/scripts/extensions/connection-manager/index.js:
```diff
@@ -549,6 +555,8 @@ async function renderDetailsContent(detailsContent) {
             }],
         });
 
+        newName = DOMPurify.sanitize(String(newName));
```
Same here. Move sanitation after if.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -146,6 +148,10 @@ function loadSettings() {
         }
     }
 
+    if (extension_settings.connectionManager) {
```
Why does this check exist? Do you need to check if the connectionManager extension is not disabled?

bmen25124 replied:
Yes

bmen25124 replied:
Oh, this is just a setting, not if the extension is enabled or not. You are right. I should change the check.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/memory/index.js:
```diff
@@ -608,6 +654,77 @@ async function getSummaryPromptForNow(context, force) {
+        return null;
+    }
+
+    console.log('sending summary prompt to connection profile');
```
Not a very useful log. Make it more informative or change level to debug.

bmen25124 replied:
Removed

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +270,300 @@ export async function getWebLlmContextSize() {
+        const { extractData, includePreset } = { ...this.defaultSendRequestParams, ...custom };
+
+        const context = SillyTavern.getContext();
+        if (!context.extensionSettings.connectionManager) {
```
Not a valid check. Please check if the extension is not disabled instead.

bmen25124 replied:
Fixed

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +270,300 @@ export async function getWebLlmContextSize() {
+            throw new Error('Connection Manager is not available');
+        }
+
+        const profile = context.extensionSettings.connectionManager.profiles.find((p) => p.id === profileId);
```
I think you can move profile validation to a private class function

bmen25124 replied:
Extracted `validateProfile` method

Cohee1207 commented on public/scripts/extensions/shared.js:
```diff
@@ -269,3 +270,300 @@ export async function getWebLlmContextSize() {
+            if (groupProfiles.length === 0) continue;
+
+            // Should we sort? I'm not sure.
+            // groupProfiles.sort((a, b) => a.name.localeCompare(b.name));
```
Sorting will be  a good idea.

bmen25124 replied:
Enabled it.

bmen25124 committed:
New validate method, better extension enable check, enabled alphabetical sort

bmen25124 committed:
Fixed name check when cancel connection manager popup

Wolfsblvt commented on public/scripts/extensions/shared.js:
```diff
@@ -402,6 +389,31 @@ export class ConnectionManagerRequestService {
+     */
+    static validateProfile(profile) {
+        if (!profile) {
+            throw new Error(`Could not find profile.`);
```
Leftover wrong quotes, linting will say to just use `'` instead of template string.

bmen25124 replied:
Oh shit, my eslint was disabled for the project. I had other errors as well. Like missing semicolon. Fixing all.

bmen25124 replied:
Fixed all.

bmen25124 committed:
Fixed eslint errors

Cohee1207 committed:
Merge branch 'staging' into memory_connection_profiles

Cohee1207 committed:
Redesign button layout

Cohee1207 committed:
Merge branch 'staging' into memory_connection_profiles

Cohee1207 reviewed (approved):
There are some happy cases when this works, but this requires more work related to sending the preset settings to the API. For example, sending a `top_k` to Groq fails with a request error. The inability to use certain settings on specifc APIs requires us to think of a solution.

bmen25124 committed:
Added presetToGeneratePayload for chat completion

bmen25124 committed:
Added presetToGeneratePayload for text completion

Cohee1207 commented on public/scripts/custom-request.js:
```diff
@@ -113,10 +122,250 @@ export class TextCompletionService {
+        const settings = { ...preset, ...customParams };
+
+        // Extract API type and determine if dynamic temperature is supported
+        const apiType = settings.api_type || textgen_types.KOBOLD;
```
I don't think that fallback type exists. It can be KOBOLDCPP

bmen25124 replied:
I'm changing this method anyway. It should look like `getTextGenGenerationData` in my opinion

bmen25124 committed:
Better presetToGeneratePayload for text completion

bmen25124 committed:
Fixed sampler_priority for APHRODITE

bmen25124 committed:
Removed negative_prompt

bmen25124 committed:
Removed stop

bmen25124 commented:
Chat completion is easy because merging `preset+customPreset` already gives us what we want.

But text completion is different. So I added`TextCompletionRequestBase` argument. Even with this, I would say this method is experimental.  I'm sure it is going to work flawlessly for 97% of the time. But:
- No `stop/stopping_strings`. Original method using `getStoppingStrings()`
- No `logprob/n_probs`. Both depend on `power_user.request_token_probabilities`. I'm not sure we should add or not.
- Banned tokens are only handling quote and number arrays. (They are not using API to tokenize.)
- No `negative_prompt`. Because the original method uses [substituteParams](https://github.com/SillyTavern/SillyTavern/blob/7de516e5e70bdb9ecff9ddb586635834c844c226/public/scripts/textgen-settings.js#L1349).
- No `dry_sequence_breakers`. Because the original method uses [replaceMacrosInList](https://github.com/SillyTavern/SillyTavern/blob/7de516e5e70bdb9ecff9ddb586635834c844c226/public/scripts/textgen-settings.js#L1307)
- No `num_ctx/truncation_length`. I don't think that we should depend on global [max_context](https://github.com/SillyTavern/SillyTavern/blob/7de516e5e70bdb9ecff9ddb586635834c844c226/public/script.js#L876) variable.



In chat completion, I removed stop and logprobs fields for the same reason. Removed tool. Removed scale ai because there was a new http request.

I highly recommend using LLMs to compare `presetToGeneratePayload` methods with [getTextGenGenerationData](https://github.com/SillyTavern/SillyTavern/blob/7de516e5e70bdb9ecff9ddb586635834c844c226/public/scripts/textgen-settings.js#L1268) and [sendOpenAIRequest](https://github.com/SillyTavern/SillyTavern/blob/7de516e5e70bdb9ecff9ddb586635834c844c226/public/scripts/openai.js#L1929).

bmen25124 committed:
Removed dry_sequence_breakers

bmen25124 committed:
Removed stream and bypass_status_check

bmen25124 commented:
Doc: https://github.com/SillyTavern/SillyTavern-Docs/pull/133

Reithan committed:
update regex depth setting tooltips for accuracy

Reithan committed:
update max tooltip

bmen25124 committed:
Merge branch 'staging' into memory_connection_profiles

Reithan committed:
remove redundant check

bmen25124 committed:
Merge branch 'staging' into memory_connection_profiles

bmen25124 committed:
presetToGeneratePayload methods are simplified

bmen25124 commented:
I simplified `presetToGeneratePayload` methods to only support `temperature/min_p` due to the next release.

Cohee1207 committed:
Merge branch 'staging' into dont-count-continue-as-message-0

Cohee1207 committed:
Allow -1 as a min depth value

Cohee1207 committed:
Merge branch 'dont-count-continue-as-message-0' of https://github.com/Reithan/SillyTavern; branch 'staging' of https://github.com/SillyTavern/SillyTavern into staging

Cohee1207 committed:
Merge branch 'staging' into memory_connection_profiles

bmen25124 committed:
custom-request progress
- Added reasoning extraction
- Added min_p to text completion

bmen25124 committed:
Added instruct format

bmen25124 committed:
Moved instruct formatting to custom-request

bmen25124 committed:
Better param handlind + eslint

bmen25124 committed:
Fixed unused field, better jsdoc

bmen25124 committed:
Merge branch 'staging' into memory_connection_profiles

bmen25124 commented:
Ready to review.

bmen25124 committed:
Added ignoreInstruct

bmen25124 committed:
Merge branch 'staging' into memory_connection_profiles

bmen25124 committed:
Merge branch 'memory_connection_profiles' of github.com:bmen25124/SillyTavern into memory_connection_profiles

bmen25124 committed:
Added ignoreInstruct to ConnectionManagerRequestService.sendRequest

bmen25124 committed:
eslint

bmen25124 committed:
Memory extension reverted.

bmen25124 commented:
Since I can't test every provider because some of them have a monthly subscription, I reverted `memory` changes. Now there is only significance `custom-request` and `shared.js` changes.

Cohee1207 commented on public/scripts/textgen-settings.js:
```diff
@@ -86,7 +86,7 @@ const OOBA_DEFAULT_ORDER = [
     'no_repeat_ngram',
 ];
-const APHRODITE_DEFAULT_ORDER = [
+export const APHRODITE_DEFAULT_ORDER = [
```
Why is it exported?

bmen25124 replied:
We are not using but we are going to use when we have proper `presetToPayload` methods. Do you want me to remove anyway? I keep exported incase of extensions going to write their own `presetToPayload`

Cohee1207 replied:
that export is very selective. Why only Aphrodite?

bmen25124 replied:
Others are already exported. Check previous [commit](https://github.com/SillyTavern/SillyTavern/blob/abf00ee0ae53138ac74debde6579487cdc774273/public/scripts/custom-request.js#L290)

Cohee1207 replied:
ooba and llamacpp orders are not exported

bmen25124 replied:
They are not using by `getTextGenGenerationData()`

bmen25124 committed:
Added "stop" and "stopping_strings" for text completion

Cohee1207 committed:
Add type for customInstruct

Cohee1207 committed:
Fix type

Cohee1207 reviewed (approved):
I tested just the impact on core code. Didn't touch anything related to custom-request.
